aopy/preproc.py

```diff
@@ -1006,25 +1024,54 @@ def _prepare_bmi3d_v0(data, metadata, max_missing_markers=10):
         elif len(sync_clock) > len(internal_clock):
             raise RuntimeError("Extra timestamps detected, something has gone horribly wrong.")
 
-        # Find the (correct) timestamps for each cycle of bmi3d's state machine
-        measured_clock = data['measure_clock_offline']
-        corrected_clock = approx_clock.copy()
-        search_radius = 0.01
-        corrected_timestamps, uncorrected_timestamps = get_measured_frame_timestamps(
-            approx_clock['timestamp'], measured_clock['timestamp'], 
-            latency_estimate, search_radius)
-        metadata['latency_measured'] = np.nanmean(corrected_timestamps - uncorrected_timestamps) - latency_estimate
-        metadata['n_missing_markers'] = np.count_nonzero(np.isnan(uncorrected_timestamps))
-
-        # Abort if there are too many missing markers
-        metadata['n_consecutive_missing_markers'] = max_repeated_nans(uncorrected_timestamps)
-        if metadata['n_consecutive_missing_markers'] < max_missing_markers:
+    # Correct the clock
+    corrected_clock = approx_clock.copy()
+    corrected_clock = rfn.append_fields(corrected_clock, 'timestamp_bmi3d', approx_clock['timestamp'], dtypes='f8')
+    if 'sync_clock' in data and len(data['sync_clock']) > 0:
+        sync_latency_estimate = 0
+        sync_search_radius = 0.01
+        timestamp_sync = get_measured_frame_timestamps(
+            approx_clock['timestamp'], data['sync_clock']['timestamp'], 
+                sync_latency_estimate, sync_search_radius)
+        corrected_clock = rfn.append_fields(corrected_clock, 'timestamp_sync', timestamp_sync, dtypes='f8')
+
+    measure_search_radius = 0.01
+    max_missing_markers = len(corrected_clock) * 0.05 # maximum 5 percent missing
```

Comment:
What is a marker?

Reply:
Also, should the 5% threshold be a input parameter?

Reply:
yes, a lot of things should be input parameters. when this is more complete i'd like to turn each big block of code into its own function. for now i think it will get too messy...